### PR TITLE
During bootup, for existing applications with RW images, avoid waiting for verification of downloaded images.

### PR DESCRIPTION
**Pull request by kalyan-nidumolu**
During bootup, for existing applications with RW images, avoid waiting for verification of downloaded images.

1) DomainMgr should publish AppUUID for the images in ImageStatus
2) Zedmanager, when creating an App instance, should check if
    the image is already available in ImageStatus for RW images.
    if yes, it should just add Verifier config and skip waiting for
    verifier
3) Verifier - Skip verifying images on bootup
    - These can be verified with the first verifier config is added.

Signed-off-by: Kalyan Nidumolu <kalyan@zadeda.com>

Fix Verifier to not wait for images in verified/ dir to be re-verified
after bootup.

Signed-off-by: Kalyan Nidumolu <kalyan@zadeda.com>

### CODE REVIEW

**Comment by kalyan-nidumolu** [comment]
Am still testing the fix. Please review the fix.

File: pkg/pillar/cmd/domainmgr/domainmgr.go [hunk 1]
```diff
@@ -351,14 +368,24 @@ func populateInitialImageStatus(ctx *domainContext, dirName string) {
...
-		log.Debugf("populateInitialImageStatus: Processing %d Mbytes %s \n",
-			size/(1024*1024), filelocation)
+		_, sha256, appUuidStr := parseAppRwImageName(filelocation)
+		log.Debugf("populateInitialImageStatus: Processing AppUuid: %s, "+
+			"%d Mbytes, fileLocation:%s",
+			appUuidStr, size/(1024*1024), filelocation)
+
+		appUUID, err := uuid.FromString(appUuidStr)
+		if err != nil {
+			log.Errorf("populateInitialImageStatus: Invalid UUIDStr%s). err: %s,",
```
Comment (eriknordmark) [inline pkg/pillar/cmd/domainmgr/domainmgr.go]: What appUUID are you publishing in ImageStatus in this case? Either force it to a known value like the all-zeros UUID, or don't publish ImageStatus?

File: pkg/pillar/cmd/verifier/verifier.go [hunk 2]
```diff
@@ -201,28 +201,7 @@ func Run() {
...
-			agentlog.StillRunning(agentName, warningTime, errorTime)
-		}
-	}
-	doneChan := make(chan struct{})
-	go keepRunning(doneChan)
-
-	handleInitUpdateVerifiedObjects(&ctx)
-
-	doneChan <- struct{}{}
+	// handleInitUpdateVerifiedObjects(&ctx)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/verifier/verifier.go]: That doesn't look like a comment, nor like code ;-)

**Comment by eriknordmark** [comment]
> 3\. Verifier - Skip verifying images on bootup
>     
>     * These can be verified with the first verifier config is added.

Doesn't zedmanager look for a VerifyImageStatus (after it has looked for a RW image), and if none found it creates a DownloaderConfig? If so change #3 will result in a duplicate/unneeded download.

File: pkg/pillar/cmd/domainmgr/domainmgr.go (hunk 1 above)
Comment (kalyan-nidumolu) [reply pkg/pillar/cmd/domainmgr/domainmgr.go]: The APP UUID is encoded in ihe image location - isn't it? I am extracting the App UUID from the image location.. A couple of lines above.. This is the Right App UUID - isn't it?

File: pkg/pillar/cmd/verifier/verifier.go (hunk 2 above)
Comment (kalyan-nidumolu) [reply pkg/pillar/cmd/verifier/verifier.go]: Will remove these functions..

**Comment by kalyan-nidumolu** [comment]
> I don't understand the design.
> zedmanager doesn't need to wait for a verified image when a RW image already exists. That's OK
> 
> But you also seem to be removing the code in verifier which re-runs the verification and publication of the verified images on boot. I don't understand why (assuming that is what you are removing; didn't complete the review due to my lack of understanding of what you are trying to do)


Here is the design - basically what we discussed:

1) Verifier - on bootup - published VerifiedImageStatus with state set to Download.
2) When Zedmanager reacts to VerifiedImageStatus, it adds VerifierImageConfig
     -- Existing Code - No changes here
3) Verifier reacts to VerifierImageConfig ( sees it as Modify ) - Does the Actual verification here..
4) Zedmanager - When creating an app Instance - checks if the imageStatus already exists for the APP UUID - if yes - directly uses it. If not - Falls through the existing code path..

This is the gist of the design..

**Comment by kalyan-nidumolu** [comment]
> > 3. Verifier - Skip verifying images on bootup
> > ```
> > * These can be verified with the first verifier config is added.
> > ```
> 
> Doesn't zedmanager look for a VerifyImageStatus (after it has looked for a RW image), and if none found it creates a DownloaderConfig? If so change #3 will result in a duplicate/unneeded download.

No..Because currently, we check for VerifierImageStatus in zedmanager..  And verifier - during bootup - publishes VerifyImageStatus with State set too DOWNLOADED. this moved to DELIVERED ( practivally immediately as explained above ) .. But irrespective on when this move, the VerifyImageStatus ensures the image doesn't gett downloaded again..

**Comment by eriknordmark** [comment]
> * Verifier reacts to VerifierImageConfig ( sees it as Modify ) - Does the Actual verification here..

That requires new code paths AFAIK.
When verifier handles a VerifyImageConfig it assumes the file is in /persist/downloaded/pending (and moves it via the verifying to the verified directories). If you want it to look in the verified directory and run the verification there you need that logic. Otherwise things are likely to break with failures about /persist/download/pending/<something> missing.

**Comment by kalyan-nidumolu** [comment]
> > * Verifier reacts to VerifierImageConfig ( sees it as Modify ) - Does the Actual verification here..
> 
> That requires new code paths AFAIK.
> When verifier handles a VerifyImageConfig it assumes the file is in /persist/downloaded/pending (and moves it via the verifying to the verified directories). If you want it to look in the verified directory and run the verification there you need that logic. Otherwise things are likely to break with failures about /persist/download/pending/ missing.

I handled it in verifier.handleModify -- The code that was being executed during init - that just takes the image in the verified directory - and verifies it - I added similar code in handleModify. So, don't need to move it through pending/verifier/verified again..

File: pkg/pillar/cmd/domainmgr/domainmgr.go (hunk 1 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/domainmgr/domainmgr.go]: My point is that you to a log.Errorf and then drive on as nothing happened. I don't know what thge return value is from uuid.FromString when it returns an error. Could be random garbage.

File: pkg/pillar/cmd/verifier/verifier.go (hunk 2 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/verifier/verifier.go]: Please remove the functions which you commented as dead as well.

File: pkg/pillar/cmd/verifier/verifier.go [hunk 3]
```diff
@@ -1158,6 +1140,31 @@ func handleModify(ctx *verifierContext, config *types.VerifyImageConfig,
 			status.Safename)
 	}
 
+	if !status.IsContainer && status.State == types.DOWNLOADED {
```
Comment (eriknordmark) [inline pkg/pillar/cmd/verifier/verifier.go]: What about the case when IsContainer is set?
Should the code be
 if DOWNLOADED {
     if isContainer {
          error
     }
     ...

Driving on when isContainer is set seems like a potential source of (future) problems.

File: pkg/pillar/cmd/zedmanager/zedmanager.go [hunk 4]
```diff
@@ -212,6 +213,20 @@ func Run() {
...
+	subImageStatus, err := pubsub.Subscribe("domainmgr",
+		types.ImageStatus{}, false, &ctx)
+	if err != nil {
+		log.Fatal(err)
+	}
+	subImageStatus.MaxProcessTimeWarn = warningTime
+	subImageStatus.MaxProcessTimeError = errorTime
+	subImageStatus.ModifyHandler = handleImageStatusCreate
+	subImageStatus.CreateHandler = handleImageStatusModify
+	subImageStatus.DeleteHandler = handleImageStatusDelete
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedmanager/zedmanager.go]: You don't need these empty handlers.

File: pkg/pillar/types/verifiertypes.go [hunk 5]
```diff
@@ -119,7 +119,8 @@ func (status VerifyImageStatus) ImageDownloadFilenames() (string, string, string
 		// Else..VMs
 		pendingFilename = pendingDirname + "/" + status.Safename
 		verifierFilename = verifierDirname + "/" + status.Safename
-		verifiedFilename = verifiedDirname + "/" + status.Safename
+		filename := SafenameToFilename(status.Safename)
+		verifiedFilename = verifiedDirname + "/" + filename
```
Comment (eriknordmark) [inline pkg/pillar/types/verifiertypes.go]: Is this an unrelated bugfix??

File: pkg/pillar/cmd/zedmanager/handleimagestatus.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+		status := cast.CastImageStatus(item)
+		if status.AppInstUUID == appUUID && status.ImageSha256 == imageSha {
+			log.Debugf("lookupImageStatusForApp: IS found. appUUID: %s, "+
+				"imageSha: %s", appUUID.String(), imageSha)
+			return status
+		}
+	}
+	log.Debugf("lookupImageStatusForApp: Image Status NOT found. appUUID: %s, "+
+		"imageSha: %s", appUUID.String(), imageSha)
+	return types.ImageStatus{}
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedmanager/handleimagestatus.go]: Odd and results in odd errors checks in callers which compare some uuid.
Either make the function return a pointer (and nil if not found), or make it return (struct, error).

File: pkg/pillar/cmd/zedmanager/updatestatus.go [hunk 7]
```diff
@@ -670,55 +695,61 @@ func doInstall(ctx *zedmanagerContext,
...
-			ss.ErrorTime = time.Time{}
-			changed = true
-		}
-		switch vs.State {
-		case types.INITIAL:
-			// Nothing to do
-		default:
-			ss.ActiveFileLocation = types.VerifiedAppImgDirname + "/" + vs.Safename
+		if is.AppInstUUID == config.UUIDandVersion.UUID {
+			ss.ActiveFileLocation = types.VerifiedAppImgDirname + "/" + safename
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedmanager/updatestatus.go]: That's incorrect. To enable multiple domUs from the same image there is at least a domainId in the string.
Doesn't ImageStatus contain the active file location? It is the directory/file name it is reporting.

File: pkg/pillar/cmd/zedmanager/updatestatus.go [hunk 8]
```diff
@@ -670,55 +695,61 @@ func doInstall(ctx *zedmanagerContext,
...
-			ss.ErrorSource = ""
-			ss.ErrorTime = time.Time{}
-			changed = true
-		}
-		switch vs.State {
-		case types.INITIAL:
-			// Nothing to do
-		default:
-			ss.ActiveFileLocation = types.VerifiedAppImgDirname + "/" + vs.Safename
+		if is.AppInstUUID == config.UUIDandVersion.UUID {
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedmanager/updatestatus.go]: See comment about error/not-found checking.

File: pkg/pillar/cmd/zedmanager/updatestatus.go (hunk 7 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/zedmanager/updatestatus.go]: Looking at the VerifyImageStatus case below it seems like ActiveFileLocation isn't the RW file, which is odd.

File: pkg/pillar/types/verifiertypes.go [hunk 9]
```diff
@@ -119,7 +119,8 @@ func (status VerifyImageStatus) ImageDownloadFilenames() (string, string, string
 		// Else..VMs
 		pendingFilename = pendingDirname + "/" + status.Safename
 		verifierFilename = verifierDirname + "/" + status.Safename
-		verifiedFilename = verifiedDirname + "/" + status.Safename
+		filename := SafenameToFilename(status.Safename)
```
Comment (eriknordmark) [inline pkg/pillar/types/verifiertypes.go]: Is this an unrelated bug fix?

File: pkg/pillar/cmd/zedmanager/handleimagestatus.go [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+func handleImageStatusCreate(ctxArg interface{}, key string,
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedmanager/handleimagestatus.go]: Don't need these three functions unless you want the log.Infof

File: pkg/pillar/cmd/verifier/verifier.go [hunk 11]
```diff
@@ -1158,6 +1140,31 @@ func handleModify(ctx *verifierContext, config *types.VerifyImageConfig,
...
+		// Modify request, but status is in DOWNLOADED state. This happens
+		// during reboot. Compute the SHA
+		_, _, verifiedFilename := status.ImageDownloadFilenames()
+		imageHash, err := computeShaFile(verifiedFilename)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/verifier/verifier.go]: This should probably do the signature check as well, but the old code didn't so that so its OK.

File: pkg/pillar/cmd/verifier/verifier.go (hunk 3 above)
Comment (kalyan-nidumolu) [reply pkg/pillar/cmd/verifier/verifier.go]: For containers, we do not have the SHA - so we don't *yet* do any verification.. Need to revisit this with Avi's change.. Infact, should be part of Avi's change on how they validate when we boot up.

File: pkg/pillar/cmd/zedmanager/zedmanager.go (hunk 4 above)
Comment (kalyan-nidumolu) [reply pkg/pillar/cmd/zedmanager/zedmanager.go]: Hmm.. I thought these are useful for debugging .. to see when the image was deleted.. But I agree.. May be I was overthinking.. Will remove them..

File: pkg/pillar/types/verifiertypes.go (hunk 5 above)
Comment (kalyan-nidumolu) [reply pkg/pillar/types/verifiertypes.go]: I think we were just lucky we did not hit this one.. Once I removed the dedicated code in Verifier to validate during bootup, the bug was uncovered. The previous code had its own hard-coded logic for the verified image location. Now the bug in this routine got exposed.

File: pkg/pillar/cmd/zedmanager/handleimagestatus.go (hunk 6 above)
Comment (kalyan-nidumolu) [reply pkg/pillar/cmd/zedmanager/handleimagestatus.go]: Hmm.. Makes sense.. Done..

File: pkg/pillar/types/verifiertypes.go (hunk 9 above)
Comment (kalyan-nidumolu) [reply pkg/pillar/types/verifiertypes.go]: See abive comment.

File: pkg/pillar/cmd/zedmanager/handleimagestatus.go (hunk 10 above)
Comment (kalyan-nidumolu) [reply pkg/pillar/cmd/zedmanager/handleimagestatus.go]: Wanted to log.Infof .. But removed them.. I think I can correlate from domain mgr logs for debugging.

File: pkg/pillar/cmd/zedmanager/updatestatus.go (hunk 7 above)
Comment (kalyan-nidumolu) [reply pkg/pillar/cmd/zedmanager/updatestatus.go]: Yeah.. This part.. I just copied from below.. vs.Safename is same as ss.Safename. Hence retained it..

File: pkg/pillar/cmd/zedmanager/updatestatus.go (hunk 8 above)
Comment (kalyan-nidumolu) [reply pkg/pillar/cmd/zedmanager/updatestatus.go]: Done.. fixed.

File: pkg/pillar/cmd/verifier/verifier.go (hunk 3 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/verifier/verifier.go]: That's not my point.

The issue is if you ever get IsContainer && types.DOWNLOADED, what should you do?
I assert you should raise an error there and then, and not treat it as a no-op.

File: pkg/pillar/cmd/zedmanager/updatestatus.go [hunk 12]
```diff
@@ -372,75 +372,90 @@ func doUpdate(ctx *zedmanagerContext,
...
-	if vs.State != ssPtr.State {
-		ssPtr.State = vs.State
-		ssPtr.Progress = 100
-		changed = true
-	}
 	log.Debugf("Verifier Status Exists for StorageEntry: Name: %s, "+
 		"ImageID: %s,  vs.RefCount: %d, ssPtr.State: %d",
 		ssPtr.Name, ssPtr.ImageID, vs.RefCount, ssPtr.State)
-	return vs, changed
+	return types.ImageStatus{}, vs, changed
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedmanager/updatestatus.go]: Again return nil.

File: pkg/pillar/cmd/zedmanager/updatestatus.go [hunk 13]
```diff
@@ -536,9 +551,19 @@ func doInstall(ctx *zedmanagerContext,
 			ss.Name, safename, ss.ImageSha256)
 
 		// Check if VerifierStatus already exists.
-		vs, statusUpdated := doInstallProcessStorageEntriesWithVerifiedImage(
+		is, vs, statusUpdated := doInstallProcessStorageEntriesWithVerifiedImage(
 			ctx, status.UUIDandVersion.UUID, ss)
 		changed = changed || statusUpdated
+		if is.AppInstUUID == status.UUIDandVersion.UUID {
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedmanager/updatestatus.go]: Again check for is == nil.

File: pkg/pillar/cmd/zedmanager/updatestatus.go (hunk 7 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/zedmanager/updatestatus.go]: OK

File: pkg/pillar/types/verifiertypes.go (hunk 5 above)
Comment (eriknordmark) [reply pkg/pillar/types/verifiertypes.go]: OK

File: pkg/pillar/cmd/zedmanager/updatestatus.go (hunk 12 above)
Comment (kalyan-nidumolu) [reply pkg/pillar/cmd/zedmanager/updatestatus.go]: Missed these.. Fixed

File: pkg/pillar/cmd/zedmanager/updatestatus.go (hunk 13 above)
Comment (kalyan-nidumolu) [reply pkg/pillar/cmd/zedmanager/updatestatus.go]: Fixed..

File: pkg/pillar/cmd/verifier/verifier.go (hunk 3 above)
Comment (kalyan-nidumolu) [reply pkg/pillar/cmd/verifier/verifier.go]: I can log an error here.. But not sure what else to do there.

"not treat it as a no-op." --> We should never be in that state. Not sure if there is a point in adding any Error handling code here. The App Instance - if we ever hit this path because of a bug - will get stuck in the error state -- Or in the init state..

I will change this to add an error..

File: pkg/pillar/cmd/domainmgr/domainmgr.go (hunk 1 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/domainmgr/domainmgr.go]: Any update?

File: pkg/pillar/cmd/verifier/verifier.go [hunk 14]
```diff
@@ -334,6 +313,7 @@ func handleInitVerifiedObjects(ctx *verifierContext) {
 }
 
 // Verify the sha/signatures and then mark as types.DELIVERED
+// TODO - THIS IS NO LONGER USED. CAN BE DELETED.
```
Comment (eriknordmark) [inline pkg/pillar/cmd/verifier/verifier.go]: Please remove dead function

File: pkg/pillar/cmd/verifier/verifier.go [hunk 15]
```diff
@@ -465,6 +445,8 @@ func populateInitialStatusFromVerified(ctx *verifierContext,
 	}
 }
 
+// updateInitialStatusFromVerified - TODO - THIS IS NO LONGER USED.
+//  CAN BE DELETED.
```
Comment (eriknordmark) [inline pkg/pillar/cmd/verifier/verifier.go]: Please remove dead function

File: pkg/pillar/cmd/verifier/verifier.go (hunk 3 above)
Comment (kalyan-nidumolu) [reply pkg/pillar/cmd/verifier/verifier.go]: Good Catch.. Actually there is a case where Container image is still in Downloader state. And we need to mark it as DELIVERED. Fixed it.. Good catch..

File: pkg/pillar/cmd/verifier/verifier.go (hunk 14 above)
Comment (kalyan-nidumolu) [reply pkg/pillar/cmd/verifier/verifier.go]: Done..

File: pkg/pillar/cmd/verifier/verifier.go (hunk 15 above)
Comment (kalyan-nidumolu) [reply pkg/pillar/cmd/verifier/verifier.go]: Done..

File: pkg/pillar/cmd/verifier/verifier.go (hunk 3 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/verifier/verifier.go]: Yeah Avi (and I) just ran into that bug in his code today! I don't understand how it worked in current master.

File: pkg/pillar/cmd/domainmgr/domainmgr.go (hunk 1 above)
Comment (kalyan-nidumolu) [reply pkg/pillar/cmd/domainmgr/domainmgr.go]: Fixed error handling.. In case of such unexpected errors, delete the file and do not publish ImageStatus
